### PR TITLE
[PyTorch] Add test for all-masked case for native softmax

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75493
* #75347
* #75346
* #75491

It returns all NaNs. CUDA implementation required a fix for this.

Differential Revision: [D35327730](https://our.internmc.facebook.com/intern/diff/D35327730/)

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -16031,6 +16031,12 @@ def embedding_bag_check(indices, weights, mode, sparse, padding_idx):
                     rtol = None
                 self.assertEqual(grad, grad_check, msg=msg, atol=atol, rtol=rtol)
 
+    def _slow_masked_softmax(self, input, mask):
```
Comment (cpuhrsch) [inline test/test_nn.py:16034]: This gives you awful numerical stability. Why not something like
```
def masked_softmax(input, dim, _stacklevel=3, dtype=None, mask=None):
    return torch.nn.functional.softmax(input.masked_fill(~mask, float('-inf')),                  
                                       dim, _stacklevel, dtype)
```

Comment (swolchok) [reply test/test_nn.py:16149]: because masked_softmax_cuda will do just that for some inputs: https://github.com/pytorch/pytorch/blob/35d4a805ebc3b6eca1bafb2d332dffa8d0c1fc54/aten/src/ATen/native/cuda/SoftMax.cu#L976

Comment (swolchok) [reply test/test_nn.py:16149]: also, this is not new code; it is a move of a function that was already present. (see line 16162 below)

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 2]
```diff
@@ -183,16 +188,6 @@ __global__ void softmax_warp_forward(output_t *dst, const input_t *src, int batc
         for (int it = 0;  it < WARP_ITERATIONS;  ++it) {
             int element_index = local_idx + it * WARP_SIZE;
             if (element_index < element_count) {
-                if (is_masked) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh:186]: So this code was just entirely unused or is it a change that still makes the tests pass?

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:186]: this path forcibly wrote zeros to output values. instead of doing that, we update the temporary `elements` array, and then let line 194 (previously 199) below produce a NaN if the entire row was masked.

Comment (cpuhrsch) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:186]: This is fine if NaNs don't cause performance regressions. We can expect to see users mask out entire rows. @ngimel can I hand this off to you?

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:186]: NaNs are the required behavior for compatibility with the existing softmax implementation.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 3]
```diff
@@ -167,6 +167,11 @@ __global__ void softmax_warp_forward(output_t *dst, const input_t *src, int batc
                         elements[i][it] = std::exp(elements[i][it] - max_value[i]);
                         sum[i] += elements[i][it];
                     }
+                } else {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: do we need this else? `elements` (masked and unmasked) are correctly initialized, so previous branch in line 163 should handle them correctly (as long as max_value is finite), we should just remove `if (!mask[idx])`

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: oh I see, we're only "saving" the addition of 0 to the sum and that's not a savings on a GPU because the branch is expensive. I'll give that a shot

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: This causes the not-masked-everywhere tests to fail due to 100% mismatch rate.

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: Would probably still need a branch protecting the addition to sum[i], at which point there's no point, right?

Comment (swolchok) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: > elements (masked and unmasked) are correctly initialized

they're not necessarily initialized to -infinity if they correspond to a masked index, so I think that's why this didn't work.

> Would probably still need a branch protecting the addition to sum[i], at which point there's no point, right?

That didn't help either, probably because the initialization is not magically to -infinity for masked locations.

Comment (ngimel) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh:170]: Hm, I thought they are initialized to -inf https://github.com/pytorch/pytorch/pull/75493/files#diff-1dc91d02c649bc32c97da8214879db0fbbd8ed3e342ba7cb6e7de9bb65b79b47R111, but apparently something is not working.

**Comment by swolchok** [comment]
ghexport stopped wanting to update this, so I've had to send a new PR -- https://github.com/pytorch/pytorch/pull/75803
